spec/rspec/matchers/built_in/start_and_end_with_spec.rb

```diff
@@ -102,11 +102,25 @@ def ==(other)
   end
 
   context "with an object that does not respond to :[]" do
-    it "fails with a useful message" do
-      actual = Object.new
-      expect {
-        expect(actual).to start_with 0
-      }.to fail_with("expected #{actual.inspect} to start with 0, but it cannot be indexed using #[]")
+    context "with an object that responds to start_with?" do
+      it "relies on start_with?" do
+        my_struct = Struct.new(:foo) do
+          def start_with?(elem)
+            true
+          end
+        end
+
+        expect(my_struct.new("foo")).to start_with(0)
+      end
+    end
+
+    context "with an object that does not respond to start_with?" do
+      it "fails with a useful message" do
+        actual = Object.new
+        expect {
+          expect(actual).to start_with 0
+        }.to fail_with("expected #{actual.inspect} to start with 0, but it does not respond to start_with? and cannot be indexed using #[]")
+      end
```

Comment:
Rather than removing the old spec and nesting additional contexts, these should be seperate cases with the original left as is

Reply:
Oh I'd love to hear more about the rationale for this!  I often set up my tests to be mutually exclusive and nesting makes that easier for me.  So they often have this shape:

```
when A
  when B
  when not B
when not A
  when B
  when not B
```
I wasn't totally sure about leaving this case as is because now, simply knowing that `actual` responds to `:[]` doesn't fully specify what should happen.  If `actual` doesn't respond to `start_with?`, it should have one behavior. If it does, it should have a different behavior.

Could you tell me more about how you were thinking of structuring this given the above?